### PR TITLE
RWA-574: Task initiation Configuration and autoassignment

### PR DESCRIPTION
**Pull request by alexbaeza**
### Change description ###

Initial implementation for RWA-574

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java [hunk 1]
```diff
@@ -118,7 +118,7 @@ public void should_return_a_200_and_retrieve_a_task_by_id_jurisdiction_location_
             .body("task.warnings", is(false));
 
 
-        common.cleanUpTask(taskId, REASON_COMPLETED);
+        //common.cleanUpTask(taskId, REASON_COMPLETED);
```
Comment (indvamshi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:121]: uncomment.

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java [hunk 2]
```diff
@@ -63,6 +88,43 @@ public RoleAssignmentResource performSearch(MultipleQueryRequest multipleQueryRe
...
+    }
+
+    private List<Classification> evaluateEqualOrHigherClassification(SecurityClassification securityClassification) {
+        switch (securityClassification) {
+            case PUBLIC:
+                return asList(PUBLIC, PRIVATE, RESTRICTED);
+            case PRIVATE:
+                return asList(PRIVATE, RESTRICTED);
+            case RESTRICTED:
+                return singletonList(RESTRICTED);
```
Comment (indvamshi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java:122]: public -> public,
private -> public, private
restricted -> public, private,restricted ????

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java (hunk 1 above)
Comment (alexbaeza) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/controllers/GetTaskByIdControllerTest.java:121]:
```suggestion
        common.cleanUpTask(taskId, REASON_COMPLETED);
```

**Review by indvamshi — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java [hunk 3]
```diff
@@ -20,6 +20,10 @@ public CFTTaskDatabaseService(TaskResourceRepository tasksRepository) {
         return tasksRepository.findById(taskId);
     }
 
+    public Optional<TaskResource> findByIdOnly(String taskId) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskDatabaseService.java:23]: Can we include coverage for this?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java [hunk 4]
```diff
@@ -346,6 +347,20 @@ public CamundaTask performGetCamundaTaskAction(String id) {
...
+    /**
+     * Updates the cftTaskState in camunda.
+     *
+     * @param taskId    the task id
+     * @param taskState the new task state
+     */
+    public void updateCftTaskState(String taskId, TaskState taskState) {
+        try {
+            updateCftTaskStateTo(taskId, taskState);
+        } catch (FeignException ex) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java:359]: could we add test to cover the exception scenario?

**Review by BenNzewi — COMMENTED** [review]
CamundaService -> performClaimTaskAction method
Can we add coverage for this;
 switch (camundaException.getType()) {
                case "TaskAlreadyClaimedException":
                    throw new ConflictException(camundaException.getMessage(), ex);
                case "NullValueException":
                    throw new ResourceNotFoundException(camundaException.getMessage(), ex);
                default:
                    throw new CamundaTaskClaimException(ex);

**Review by BenNzewi — COMMENTED** [review]
CamunndaService-> could we add coverage for this;

private boolean filterOnlyHasWarningVarAndLocalTaskVars(CamundaVariableInstance camundaVariableInstance) {
        if (("hasWarnings".equals(camundaVariableInstance.getName())
             || "warningList".equals(camundaVariableInstance.getName()))
            && camundaVariableInstance.getTaskId() == null) {
            return true;
        }
        return camundaVariableInstance.getTaskId() != null;
    }

**Review by BenNzewi — COMMENTED** [review]
CamundaService - could we add coverage for 

1. private void performCompleteTaskAction(String taskId, boolean taskHasCompleted) {
        if (!taskHasCompleted) {
            // If task was not already completed complete it
            updateTaskStateTo(taskId, TaskState.COMPLETED);
        }

2. private void performUnclaimTaskAction(String taskId, boolean taskHasUnassigned) {
        if (!taskHasUnassigned) {
            updateTaskStateTo(taskId, TaskState.UNASSIGNED);
        }

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java [hunk 5]
```diff
@@ -553,6 +568,30 @@ private void performUnclaimTaskAction(String taskId, boolean taskHasUnassigned)
...
+     */
+    private void updateCftTaskStateTo(String taskId, TaskState newState) {
+        Map<String, CamundaValue<String>> variable = Map.of(
+            CFT_TASK_STATE.value(), CamundaValue.stringValue(newState.value())
+        );
+        AddLocalVariableRequest camundaLocalVariables = new AddLocalVariableRequest(variable);
+
+        try {
+            camundaServiceApi.addLocalVariablesToTask(authTokenGenerator.generate(), taskId, camundaLocalVariables);
+        } catch (FeignException ex) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CamundaService.java:584]: could we add test to cover the exception scenario?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java [hunk 6]
```diff
@@ -63,6 +88,43 @@ public RoleAssignmentResource performSearch(MultipleQueryRequest multipleQueryRe
...
 
+    private MultipleQueryRequest buildQueryForAutoAssignment(TaskResource taskResource) {
+        Set<TaskRoleResource> rolesInTask = taskResource.getTaskRoleResources();
+
+        String caseId = taskResource.getCaseId();
+        List<Classification> securityClassifications =
+            evaluateEqualOrHigherClassification(taskResource.getSecurityClassification());
+
+        Set<String> roleNamesFound = rolesInTask.stream()
+            .map(role -> role.getOwn() && role.getAutoAssignable() ? role.getRoleName() : null)
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java:99]: could add coverage for this logic?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java [hunk 7]
```diff
@@ -63,6 +88,43 @@ public RoleAssignmentResource performSearch(MultipleQueryRequest multipleQueryRe
...
+            .grantType(singletonList(GrantType.SPECIFIC))
+            .validAt(LocalDateTime.now())
+            .hasAttributes(singletonList("caseId"))
+            .attributes(Map.of("caseId", List.of(caseId)))
+            .build();
+
+        return MultipleQueryRequest.builder().queryRequests(singletonList(queryRequest)).build();
+    }
+
+    private List<Classification> evaluateEqualOrHigherClassification(SecurityClassification securityClassification) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/auth/role/TaskConfigurationRoleAssignmentService.java:115]: could add coverage for this logic?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 8]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+            Objects.requireNonNull(permission.getName(), "Permissions name cannot be null");
+            Objects.requireNonNull(permission.getValue(), "Permissions value cannot be null");
+            final String roleName = permission.getName().getValue();
+            final String permissionsValue = permission.getValue().getValue();
+
+            final Set<PermissionTypes> permissionsFound = Arrays.stream(permissionsValue.split(","))
+                .map(p -> PermissionTypes.from(p).orElse(null))
+                .collect(Collectors.toSet());
+            String[] authorisations = {};
+            if (permission.getAuthorisations() != null && permission.getAuthorisations().getValue() != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:156]: could we add coverage for this logic

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 9]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+                authorisations = permission.getAuthorisations().getValue().split(",");
+            }
+
+            Integer assignmentPriority = null;
+            if (permission.getAssignmentPriority() != null && permission.getAssignmentPriority().getValue() != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:161]: could we add coverage for this logic

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 10]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+                assignmentPriority = permission.getAssignmentPriority().getValue();
+            }
+            boolean autoAssignable = false;
+            if (permission.getAutoAssignable() != null && permission.getAutoAssignable().getValue() != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:165]: could we add coverage for this logic

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 11]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+                permissionsFound.contains(PermissionTypes.CANCEL),
+                permissionsFound.contains(PermissionTypes.REFER),
+                authorisations,
+                assignmentPriority,
+                autoAssignable
+            );
+        }).collect(Collectors.toSet());
+    }
+
+    private void mapVariableToTaskResourceProperty(TaskResource taskResource, String key, Object value) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:184]: could we review and add coverage for missed cases?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 12]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+                    break;
+                case CASE_NAME:
+                    taskResource.setCaseName((String) value);
+                    break;
+                case CASE_TYPE_ID:
+                    taskResource.setCaseTypeId((String) value);
+                    break;
+                case EXECUTION_TYPE:
+                    Optional<ExecutionType> executionType = ExecutionType.from((String) value);
+                    if (executionType.isPresent()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:207]: coverage for executionType.isPresent()

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 13]
```diff
@@ -43,6 +57,85 @@ public AutoAssignmentResult getAutoAssignmentVariables(TaskToConfigure task) {
...
+        if (!roleAssignments.isEmpty()) {
+            // the lowest assignment priority takes precedence.
+            List<TaskRoleResource> rolesList = new ArrayList<>(taskResource.getTaskRoleResources());
+            rolesList.sort(comparing(TaskRoleResource::getAssignmentPriority, nullsLast(Comparator.naturalOrder())));
+
+            Map<String, TaskRoleResource> roleResourceMap = rolesList.stream()
+                .collect(toMap(
+                    TaskRoleResource::getRoleName,
+                    taskRoleResource -> taskRoleResource,
+                    (a, b) -> b
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:83]: could we add coverage?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 14]
```diff
@@ -43,6 +57,85 @@ public AutoAssignmentResult getAutoAssignmentVariables(TaskToConfigure task) {
...
+                ));
+
+            List<RoleAssignment> orderedRoleAssignments = orderRoleAssignments(rolesList, roleAssignments);
+            Optional<RoleAssignment> match = orderedRoleAssignments.stream()
+                .filter(roleAssignment -> {
+                    TaskRoleResource taskRoleResource = roleResourceMap.get(roleAssignment.getRoleName());
+
+                    if (taskRoleResource.getAuthorizations() != null
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:91]: could we add coverage for this if logic?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 15]
```diff
@@ -43,6 +57,85 @@ public AutoAssignmentResult getAutoAssignmentVariables(TaskToConfigure task) {
...
+                        && taskRoleResource.getAuthorizations().length != 0
+                        && !roleAssignment.getAuthorisations().isEmpty()) {
+                        return findMatchingRoleAssignment(taskRoleResource, roleAssignment);
+                    }
+                    return false;
+                }).findFirst();
+
+            if (match.isPresent()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:99]: could we add coverage?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 16]
```diff
@@ -43,6 +57,85 @@ public AutoAssignmentResult getAutoAssignmentVariables(TaskToConfigure task) {
...
+                //The actorId of the role assignment selected in stage above is the IdAM ID of the user who is
+                //to be assigned the task.
+                taskResource.setAssignee(match.get().getActorId());
+                taskResource.setState(CFTTaskState.ASSIGNED);
+            } else {
+                //If stage above produces an empty result of matching role assignment, then the task is
+                //to be left unassigned.
+                taskResource.setAssignee(null);
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:107]: could we add coverage for these?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java [hunk 17]
```diff
@@ -43,6 +57,85 @@ public AutoAssignmentResult getAutoAssignmentVariables(TaskToConfigure task) {
...
+
+        return taskResource;
+    }
+
+    private boolean findMatchingRoleAssignment(TaskRoleResource taskRoleResource, RoleAssignment roleAssignment) {
+        AtomicBoolean hasMatch = new AtomicBoolean(false);
+        stream(taskRoleResource.getAuthorizations())
+            .forEach(auth -> {
+                //Safe-guard
+                if (!hasMatch.get() && roleAssignment.getAuthorisations().contains(auth)) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/TaskAutoAssignmentService.java:120]: can we add coverage for this if?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/ConfigureTaskService.java [hunk 18]
```diff
@@ -76,34 +78,72 @@ public void configureTask(String taskId) {
...
+     * Helper method to combine a task configuration result into one final object
+     * containing all process variables and evaluations.
+     *
+     * @param result               the task configuration result returned by the configurator
+     * @param configurationResults the final object where values get added
+     */
+    private void combineResults(TaskConfigurationResults result,
+                                TaskConfigurationResults configurationResults) {
+
+        if (result.getProcessVariables() != null && configurationResults.getProcessVariables() != null) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/taskconfiguration/services/ConfigureTaskService.java:135]: could we add missing coverage for this logic?

File: src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java [hunk 19]
```diff
@@ -111,10 +124,154 @@ public TaskResource mapToTaskObject(String taskId, List<TaskAttribute> taskAttri
...
+                    break;
+                case TASK_ID:
+                    taskResource.setTaskId((String) value);
+                    break;
+                case TASK_NAME:
+                    taskResource.setTaskName((String) value);
+                    break;
+                case TASK_STATE:
+                    Optional<CFTTaskState> state = CFTTaskState.from((String) value);
+                    if (state.isPresent()) {
```
Comment (BenNzewi) [inline src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/CFTTaskMapper.java:241]: coverage for CFTTaskState if logic (edge cases)

**Pull request merged**